### PR TITLE
stages/disks: improve error reporting for LUKS device reuse

### PR DESCRIPTION
**Pull request by bgilbert**
Report the specific reason that we declined to reuse a device.  Propagate reuse errors earlier so we don't have to guess why reuse failed.

Fixes https://github.com/coreos/ignition/issues/1209.

### CODE REVIEW

**Review by arithx — APPROVED** [review]

File: internal/exec/stages/disks/luks.go [hunk 1]
```diff
@@ -341,33 +335,52 @@ func (s *stage) createLuks(config types.Config) error {
...
+
+// Check LUKS device against config and open it.
+func (s *stage) reuseLuksDevice(luks types.Luks, keyFilePath string) error {
+	devAlias := execUtil.DeviceAlias(*luks.Device)
+
+	// don't allow clevis device re-use as we can't guarantee that what
+	// is stored in the header is exactly what would be generated from
+	// the given config
+	if luks.Clevis.IsPresent() {
+		return fmt.Errorf("clevis must not be configured")
```
Comment (rugk) [inline internal/exec/stages/disks/luks.go:358]: The message does not make it clear _where_ it must not be configured IMHO… In the ignition config (where it is also somewhat “configured”) or in the LUKS device?

AFAIK it’s the volume here…

So maybe:
```suggestion
		return fmt.Errorf("clevis must not be configured for the provided volume")
```

Or with device:
```suggestion
		return fmt.Errorf("clevis must not be configured for the provided device")
```

File: internal/exec/stages/disks/luks.go [hunk 2]
```diff
@@ -341,33 +335,52 @@ func (s *stage) createLuks(config types.Config) error {
...
+	}
+
+	// ephemeral keyfiles won't match the existing device
+	if util.NilOrEmpty(luks.KeyFile.Source) {
+		return fmt.Errorf("keyfile must be configured")
```
Comment (rugk) [inline internal/exec/stages/disks/luks.go:363]: Same here.

```suggestion
		return fmt.Errorf("keyfile must be configured for the provided volume")
```

or:

```suggestion
		return fmt.Errorf("keyfile must be configured for the provided device")
```

File: internal/exec/stages/disks/luks.go [hunk 3]
```diff
@@ -341,33 +335,52 @@ func (s *stage) createLuks(config types.Config) error {
...
 
 	if luks.Label != nil {
 		fsInfo, err := execUtil.GetFilesystemInfo(devAlias, true)
 		if err != nil {
-			return false, fmt.Errorf("retrieving filesystem info: %v", err)
+			return fmt.Errorf("retrieving filesystem info: %v", err)
 		}
 		if fsInfo.Label != *luks.Label {
-			return false, nil
+			return fmt.Errorf("configured label %q doesn't match volume label %q", *luks.Label, fsInfo.Label)
```
Comment (rugk) [inline internal/exec/stages/disks/luks.go:372]:
```suggestion
			return fmt.Errorf("configured label %q doesn't match the expected volume label %q", *luks.Label, fsInfo.Label)
```

Maybe just phrase it like “unit tests” do, to make it sure what is _expected_ and what is the _actual_ thing we’ve read.

File: internal/exec/stages/disks/luks.go [hunk 4]
```diff
@@ -341,33 +335,52 @@ func (s *stage) createLuks(config types.Config) error {
...
 		uuid, err := exec.Command(distro.CryptsetupCmd(), "luksUUID", devAlias).CombinedOutput()
 		if err != nil {
-			return false, err
+			return err
 		}
-		if strings.TrimSpace(string(uuid)) != *luks.UUID {
-			return false, nil
+		uuidStr := strings.TrimSpace(string(uuid))
+		if uuidStr != *luks.UUID {
+			return fmt.Errorf("configured UUID %q doesn't match volume UUID %q", *luks.UUID, uuidStr)
```
Comment (rugk) [inline internal/exec/stages/disks/luks.go:383]:
```suggestion
			return fmt.Errorf("configured UUID %q doesn't match the expected volume UUID %q", *luks.UUID, uuidStr)
```

(same here)

**Review by rugk — CHANGES_REQUESTED** [review]
Some suggestions to make the error messages potentially more clear and avoid misunderstandings… (based on my experience)

**Comment by bgilbert** [comment]
Thanks for the feedback.  Updated.

**Review by rugk — APPROVED** [review]
Thanks! Looks way clearer to me now and would have saved me a lot of debugging I guess. :see_no_evil:

**Pull request merged**